### PR TITLE
GitHub Configuration

### PR DESCRIPTION
**Pull request by nturinski**
Highly anticipated GitHub integration!

As discussed, entry point is through "Change Deployment Source" now.  This required a change in the function definition in `editScmType()` that will have to be updated in both Azure App Service and Azure Functions.

### CODE REVIEW

File: appservice/src/SiteWrapper.ts [hunk 1]
```diff
@@ -170,6 +176,61 @@ export class SiteWrapper {
         outputChannel.appendLine('');
     }
 
+    public async connectToGitHub(node: IAzureNode, outputChannel: vscode.OutputChannel): Promise<void> {
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: This can be private, right? No need to 'expose' it until necessary

File: appservice/src/SiteWrapper.ts [hunk 2]
```diff
@@ -170,6 +176,61 @@ export class SiteWrapper {
...
+            } catch (error) {
+                const parsedError: IParsedError = JSON.parse(parseError(error).message);
+                // The portal returns 200, but is expecting a 204 which causes it to throw an error even after a successful sync
+                if (parsedError.statusCode !== 200) {
+                    throw error;
+                }
+            }
+        }
+
+        this.log(outputChannel, `"${this.appName}" has been connected to GitHub repo.`);
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: Is this message redundant? Considering you already have a message at the end of editScmType

File: appservice/src/SiteWrapper.ts [hunk 3]
```diff
@@ -229,7 +300,7 @@ export class SiteWrapper {
         // tslint:disable-next-line:no-floating-promises
         actionHandler.callWithTelemetry('appService.streamingLogs', async () => {
             await new Promise((resolve: () => void, reject: (err: Error) => void): void => {
-                const logsRequest: request.Request = requestApi(`${this.kuduUrl}/api/logstream/${path}`);
+                const logsRequest: request.Request = requestApi(`${this.kuduUrl} / api / logstream / ${path}`);
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: Was this auto-formatted or something?

File: appservice/src/SiteWrapper.ts [hunk 4]
```diff
@@ -346,6 +417,75 @@ export class SiteWrapper {
...
     }
 
+    private async getJsonRequest(url: string, token: string, node: IAzureNode): Promise<Object[]> {
+        // Reference for GitHub REST routes
+        // https://developer.github.com/v3/
+        // Note: blank after user implies look up authorized user
+        try {
+            const gitHubResponse: string = await requestP.get(url, {
+                headers: {
+                    Authorization: `token ${token}`,
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: Let's use `new TokenCredentials(token)` and the signRequest method from the PR I just merged

File: appservice/src/SiteWrapper.ts [hunk 5]
```diff
@@ -201,11 +262,21 @@ export class SiteWrapper {
         return new KuduClient(cred, this.kuduUrl);
     }
 
-    public async editScmType(client: WebSiteManagementClient): Promise<string | undefined> {
+    public async editScmType(node: IAzureNode, client: WebSiteManagementClient, outputChannel: vscode.OutputChannel): Promise<string | undefined> {
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: If you're passing in `node`, there's no need to pass in `client`

File: appservice/src/SiteWrapper.ts [hunk 6]
```diff
@@ -170,6 +176,61 @@ export class SiteWrapper {
...
+        if (!oAuth2Token) {
+            this.showGitHubAuthPrompt(node);
+            return;
+        }
+
+        const gitHubUser: Object[] = await this.getJsonRequest('https://api.github.com/user', oAuth2Token, node);
+
+        const gitHubOrgs: Object[] = await this.getJsonRequest('https://api.github.com/user/orgs', oAuth2Token, node);
+        const orgQuickPicks: IQuickPickItemWithData<{}>[] = this.createQuickPickFromJsons([gitHubUser], 'login', undefined, ['repos_url']).concat(this.createQuickPickFromJsons(gitHubOrgs, 'login', undefined, ['repos_url']));
+        const orgQuickPick: gitHubOrgData = (await vscode.window.showQuickPick(orgQuickPicks, { placeHolder: 'Choose your organization.', ignoreFocusOut: true })).data;
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: Feels like you're not handling the case if the user cancels. Aka it should throw a UserCancelledError

File: appservice/src/SiteWrapper.ts [hunk 7]
```diff
@@ -170,6 +176,61 @@ export class SiteWrapper {
...
+        this.log(outputChannel, `"${this.appName}" is being connected to GitHub repo. This may take a several minutes...`);
+        try {
+            await client.webApps.createOrUpdateSourceControlWithHttpOperationResponse(this.resourceGroup, this.name, siteSourceControl);
+        } catch (err) {
+            try {
+                // a resync will fix the first broken build
+                // https://github.com/projectkudu/kudu/issues/2277
+                await client.webApps.syncRepository(this.resourceGroup, this.name);
+            } catch (error) {
+                const parsedError: IParsedError = JSON.parse(parseError(error).message);
```
[the remark this answers is not in the document]
Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: https://github.com/Microsoft/vscode-azuretools/issues/77

File: appservice/package.json [hunk 8]
```diff
@@ -38,6 +38,8 @@
         "ms-rest": "^2.2.2",
         "ms-rest-azure": "^2.4.4",
         "opn": "^5.1.0",
+        "request": "^2.83.0",
+        "request-promise": "^4.2.2",
```
[the remark this answers is not in the document]
Comment (nturinski) [reply appservice/package.json]: Registered too

File: appservice/src/SiteWrapper.ts [hunk 9]
```diff
@@ -362,12 +381,76 @@ export class SiteWrapper {
...
 
+    private async getJsonRequest(url: string, requestOptions: WebResource, node: IAzureNode): Promise<Object[]> {
+        // Reference for GitHub REST routes
+        // https://developer.github.com/v3/
+        // Note: blank after user implies look up authorized user
+        try {
+            const gitHubResponse: string = await requestP.get(url, requestOptions);
+            return JSON.parse(gitHubResponse);
+        } catch (error) {
+            if (error.message.indexOf('401 - "{\"message\":\"Bad credentials\",\"documentation_url\":\"https://developer.github.com/v3\"}"') > -1) {
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: Use `parseError(error).message`. It's safer and handles the case where message might be undefined.

Also do we need to look for that _entire_ message or can you just check for 401 or "Bad credentials"? I'm worried they might slightly change the message and then this check no longer works

File: appservice/src/SiteWrapper.ts [hunk 10]
```diff
@@ -362,12 +381,76 @@ export class SiteWrapper {
...
+                // the default error is just "Bad Credentials," which is an unhelpful error message
+                const tokenExpired: string = localize('tokenExpired', 'Azure\'s GitHub token has expired.  Reauthorize in the Portal under "Deployment options."');
+                const goToPortal: string = localize('goToPortal', 'Go to Portal');
+                const input: string | undefined = await vscode.window.showErrorMessage(tokenExpired, goToPortal);
+                if (input === goToPortal) {
+                    node.openInPortal();
+                    // this needs to be replaced with a custom error
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]: I don't like 'TODO's' as comments in code - we'll just forget about it. File an issue instead

File: appservice/thirdpartynotices.txt [hunk 11]
```diff
@@ -6,8 +6,10 @@ open source projects along with the license information below. We acknowledge an
 are grateful to these developers for their contribution to open source.
 
 1. node-archiver (https://github.com/archiverjs/node-archiver)
-2. simple-git version 1.80.1 (https://github.com/steveukx/git-js)
-3. opn version 5.1.0 (https://github.com/sindresorhus/opn
+2. simple-git (https://github.com/steveukx/git-js)
+3. opn (https://github.com/sindresorhus/opn
+4. request (https://github.com/request/request)
+5. request-promise (https://github.com/request/request-promise/blob/master/LICENSE)
```
Comment (EricJizbaMSFT) [inline appservice/thirdpartynotices.txt]: This url should go to the main repo, not the license

File: ui/index.d.ts [hunk 12]
```diff
@@ -196,4 +196,5 @@ export interface IParsedError {
     errorType: string;
     message: string;
     isUserCancelledError: boolean;
+    statusCode?: number;
```
Comment (EricJizbaMSFT) [inline ui/index.d.ts]: Undo this change - I assume you'll fix it as a part of #77

**Comment by nturinski** [comment]
There seem to be a few npm lint rules that blocking the build that include lines that I did not add.  Did we change anything in the travis.yml?

**Comment by EricJizbaMSFT** [comment]
Nope - didn't change anything. Can you clarify which rules you're talking about? I took a quick look and they mostly look like your code from this PR to me

File: appservice/src/SiteWrapper.ts (hunk 9 above)
Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: I'll use "Bad credentials".  Seems like a safe bet to me.

File: appservice/src/SiteWrapper.ts (hunk 10 above)
Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: https://github.com/Microsoft/vscode-azuretools/issues/78

**Comment by nturinski** [comment]
Specifically, line 288:
`const adminKey: string = await kuduClient.functionModel.getAdminToken();`

is also triggering `unsafe-use-any` as the Promises returns any,

**Comment by EricJizbaMSFT** [comment]
I don't see line 288 listed on [this build](https://travis-ci.org/Microsoft/vscode-azuretools/jobs/344591759)
screen shot 2018-02-21 at 5 21 40 pm

Or when I run locally from your branch:
screen shot 2018-02-21 at 5 22 39 pm

**Comment by EricJizbaMSFT** [comment]
Do you have an alternate version of the kudu package linked locally? Maybe that's why you're seeing the error...

**Comment by nturinski** [comment]
image

 😵 

I don't think so, but I'll try doing a clean install.
